### PR TITLE
OCPBUGS-2028: Set kubelet hostname for GCP platform

### PR DESCRIPTION
**Pull request by jrvaldes**
This change fetches the hostname from the GCE instance metadata service
and overrides the kubelet hostname so that the name of the Windows node
matches the hostname. The hostname is truncated if the length is longer
than 63 characters.

Should go after:
- https://github.com/openshift/windows-machine-config-operator/pull/1279

### CODE REVIEW

File: pkg/services/services.go [hunk 1]
```diff
@@ -194,12 +190,57 @@ func klogVerbosityArg(debug bool) string {
 }
 
 // getKubeletHostnameOverride returns the hostname override arg that should be used
-func getKubeletHostnameOverride(platformType string) string {
+func getKubeletHostnameOverride(platformType string) (string, servicescm.PowershellCmdArg) {
```
Comment (jrvaldes) [inline pkg/services/services.go:193]: This addresses https://github.com/openshift/windows-machine-config-operator/pull/1260#discussion_r997390403

File: pkg/services/services.go [hunk 2]
```diff
@@ -194,12 +190,57 @@ func klogVerbosityArg(debug bool) string {
...
+}
+
+// getGCPMetadataHostnamePowershellCmdArg returns the PowershellCmdArg to resolve the hostname using the instance
+// metadata service in GCP. Uses GoogleCloud Tools for PowerShell (https://cloud.google.com/tools/powershell/docs/quickstart)
+// getGCPMetadataHostname returns the GCP instance hostname from the metadata service.  GCP is an especial case,
+// and the resulting hostname is truncated if the length is longer than 63 characters.
+func getGCPMetadataHostnamePowershellCmdArg(variableName string) servicescm.PowershellCmdArg {
+	return servicescm.PowershellCmdArg{
+		Name: variableName,
+		Path: `
```
Comment (jrvaldes) [inline pkg/services/services.go:223]: As an alternative, I can move this PowerShell script into a file under the [pkg/internal](https://github.com/openshift/windows-machine-config-operator/tree/master/pkg/internal) directory, add the file to the payload and here just declare the path. Thoughts?

File: pkg/services/services.go [hunk 3]
```diff
@@ -194,12 +190,57 @@ func klogVerbosityArg(debug bool) string {
...
+# MAX_LENGTH is the maximum number of character allowed for the instance's hostname in GCP
+$MAX_LENGTH = 63
+
+# get hostname from the instance metadata service
+$hostname=Get-GceMetadata -Path "instance/hostname"
```
Comment (jrvaldes) [inline pkg/services/services.go:228]: For consistency, I followed the existing approach and am still concerned about https://github.com/openshift/windows-machine-config-operator/pull/1260#discussion_r997394197 .

I feel querying the instance metadata endpoint directly may be safer, thinking of the BYOH use case. For example in GCP:
```powershell
$HOSTNAME = (Invoke-RestMethod ` -Headers @{'Metadata-Flavor' = 'Google'} `  -Uri "http://metadata.google.internal/computeMetadata/v1/instance/hostname")
```

**Review by jrvaldes — COMMENTED** [review]
Added a [comment](https://issues.redhat.com/browse/OCPBUGS-2028?focusedCommentId=21122869&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-21122869) in the ticket capturing the machine and node outputs using the proposed logic to truncate longer hostnames.

I tested this with short and long machine names, and the resulting node name is adjusted accordingly. 

```
[windows-machine-config-operator]$ date && oc get machines -A && oc get nodes
Wed Oct 19 02:31:57 PM EDT 2022
NAMESPACE               NAME                                                              PHASE     TYPE            REGION        ZONE            AGE
openshift-machine-api   jvaldes-vcjw5-master-0                                            Running   n2-standard-4   us-central1   us-central1-a   4h14m
openshift-machine-api   jvaldes-vcjw5-master-1                                            Running   n2-standard-4   us-central1   us-central1-b   4h14m
openshift-machine-api   jvaldes-vcjw5-master-2                                            Running   n2-standard-4   us-central1   us-central1-c   4h14m
openshift-machine-api   jvaldes-vcjw5-windows-worker-with-the-max-length-in-name-6vnsk7   Running   n1-standard-4   us-central1   us-central1-a   11m
openshift-machine-api   jvaldes-vcjw5-winworker-a-ssgrf                                   Running   n1-standard-4   us-central1   us-central1-a   3h6m
openshift-machine-api   jvaldes-vcjw5-worker-a-lttk7                                      Running   n2-standard-4   us-central1   us-central1-a   4h7m
openshift-machine-api   jvaldes-vcjw5-worker-b-gxtjn                                      Running   n2-standard-4   us-central1   us-central1-b   4h7m
openshift-machine-api   jvaldes-vcjw5-worker-c-xmbsk                                      Running   n2-standard-4   us-central1   us-central1-c   4h7m
NAME                                                              STATUS   ROLES                  AGE     VERSION
jvaldes-vcjw5-master-0.c.openshift-gce-devel.internal             Ready    control-plane,master   4h11m   v1.25.0+3ef6ef3
jvaldes-vcjw5-master-1.c.openshift-gce-devel.internal             Ready    control-plane,master   4h11m   v1.25.0+3ef6ef3
jvaldes-vcjw5-master-2.c.openshift-gce-devel.internal             Ready    control-plane,master   4h11m   v1.25.0+3ef6ef3
jvaldes-vcjw5-windows-worker-with-the-max-length-in-name-6vnsk7   Ready    worker                 4m54s   v1.25.0-2492+3ef6ef3cfbd273
jvaldes-vcjw5-winworker-a-ssgrf.c.openshift-gce-devel.internal    Ready    worker                 3h      v1.25.0-2492+3ef6ef3cfbd273
jvaldes-vcjw5-worker-a-lttk7.c.openshift-gce-devel.internal       Ready    worker                 3h57m   v1.25.0+3ef6ef3
jvaldes-vcjw5-worker-b-gxtjn.c.openshift-gce-devel.internal       Ready    worker                 3h57m   v1.25.0+3ef6ef3
jvaldes-vcjw5-worker-c-xmbsk.c.openshift-gce-devel.internal       Ready    worker                 3h57m   v1.25.0+3ef6ef3

```

File: pkg/services/services.go (hunk 2 above)
Comment (saifshaikh48) [reply pkg/services/services.go:223]: I agree with the proposed alternative. Having the script live in a separate file might be a cleaner approach.

Comment (aravindhp) [reply pkg/services/services.go:222]: +1. That is is the pattern we have been following so far.

File: pkg/services/services.go (hunk 3 above)
Comment (aravindhp) [reply pkg/services/services.go:227]: Can we be sure that the endpoint won't change?

Comment (jrvaldes) [reply pkg/services/services.go:227]: I feel we can trust it. The `metadata.google.internal` DNS name points to the IPv4 address of the instance metadata service: `169.254.169.254`. 

```
PS C:\Users\administrator> cat C:\Windows\System32\drivers\etc\hosts   
# Copyright (c) 1993-2009 Microsoft Corp.
#
# This is a sample HOSTS file used by Microsoft TCP/IP for Windows.
#
# This file contains the mappings of IP addresses to host names. Each
# entry should be kept on an individual line. The IP address should
# be placed in the first column followed by the corresponding host name.
# The IP address and the host name should be separated by at least one
# space.
#
# Additionally, comments (such as these) may be inserted on individual
# lines or following the machine name denoted by a '#' symbol.
#
# For example:
#
#      102.54.94.97     rhino.acme.com          # source server
#       38.25.63.10     x.acme.com              # x client host

# localhost name resolution is handled within DNS itself.
#       127.0.0.1       localhost
#       ::1             localhost

# Google Compute Engine metadata server
    169.254.169.254    metadata.google.internal metadata

```

Comment (aravindhp) [reply pkg/services/services.go:227]: My point is if we can trust that `metadata.google.internal` itself won't change. Just because it is in a `/etc/hosts` file does not mean that the content cannot change.

Comment (jrvaldes) [reply pkg/services/services.go:227]: If we are trusting the `Get-GceMetadata` command from the GCP Cloud Tools for PowerShell, we can trust the `metadata.google.internal` endpoint as this is what the library is [using](https://github.com/GoogleCloudPlatform/google-cloud-powershell/blob/b0549f98b2a150830ff6ddf10c66c3252df7626d/Google.PowerShell/Compute/GceMetadataCmdlet.cs#L131).

File: pkg/services/services.go [hunk 4]
```diff
@@ -193,12 +189,57 @@ func klogVerbosityArg(debug bool) string {
...
+
+// getGCPMetadataHostnamePowershellCmdArg returns the PowershellCmdArg to resolve the hostname using the instance
+// metadata service in GCP. Uses GoogleCloud Tools for PowerShell (https://cloud.google.com/tools/powershell/docs/quickstart)
+// getGCPMetadataHostname returns the GCP instance hostname from the metadata service.  GCP is an especial case,
+// and the resulting hostname is truncated if the length is longer than 63 characters.
+func getGCPMetadataHostnamePowershellCmdArg(variableName string) servicescm.PowershellCmdArg {
+	return servicescm.PowershellCmdArg{
+		Name: variableName,
+		Path: `
+# MAX_LENGTH is the maximum number of character allowed for the instance's hostname in GCP
```
Comment (alinaryan) [inline pkg/services/services.go:223]: Is this/should this be doc'd somewhere so no one accidentally creates an instance more than 63 chars?

File: pkg/services/services.go (hunk 2 above)
Comment (alinaryan) [reply pkg/services/services.go:222]: +1

**Review by alinaryan — COMMENTED** [review]
Looks good so far: Just a comment

File: pkg/services/services.go (hunk 3 above)
Comment (aravindhp) [reply pkg/services/services.go:227]: Point of using a library is that it abstracts such details from us.

Comment (aravindhp) [reply pkg/services/services.go:227]: What I am looking for is a reference to a GCE doc that guarantees this endpoint won't change or a way for us to easily catch the change in an e2e test.

File: pkg/services/services.go (hunk 2 above)
Comment (aravindhp) [reply pkg/services/services.go:222]: To be clear the request is to include this as a file in the internal directory

Comment (saifshaikh48) [reply pkg/services/services.go]: You could check platform before adding it to the files to be transferred in `windows.go`. We currently transfer `azure-cloud-node-manager.exe` to every instance on every platform though, so this may warrant a team-wide consensus on whether everything in the payload should always be transferred as is.

Comment (aravindhp) [reply pkg/services/services.go]: +1

And this should not be the reason to not make this file

File: pkg/services/services.go [hunk 5]
```diff
@@ -236,29 +236,11 @@ func getAWSMetadataHostnamePowershellCmdArg(variableName string) servicescm.Powe
 }
 
 // getGCPMetadataHostnamePowershellCmdArg returns the PowershellCmdArg to resolve the hostname using the instance
-// metadata service in GCP. Uses GoogleCloud Tools for PowerShell (https://cloud.google.com/tools/powershell/docs/quickstart)
-// getGCPMetadataHostname returns the GCP instance hostname from the metadata service.  GCP is an especial case,
-// and the resulting hostname is truncated if the length is longer than 63 characters.
+// metadata service in GCP.
+// GCP is an especial case, and the resulting hostname is truncated if the length is longer than 63 characters.
```
Comment (saifshaikh48) [inline pkg/services/services.go:240]: `especial` --> `special`

File: pkg/services/services.go [hunk 6]
```diff
@@ -144,13 +143,10 @@ func getKubeletServiceConfiguration(argsFromIginition map[string]string, debug b
 		return servicescm.Service{}, err
 	}
 	var powershellVars []servicescm.PowershellCmdArg
-	hostnameArg := getKubeletHostnameOverride(platform)
+	hostnameArg, hostnamePowershellVars := getKubeletHostnameOverride(platform)
```
Comment (saifshaikh48) [inline pkg/services/services.go:146]: nit: `hostnamePowershellVar` instead of the plural `hostnamePowershellVars`, I had to re-read to make sure `getKubeletHostnameOverride()` only return one PowershellCmdArg

**Review by saifshaikh48 — COMMENTED** [review]
My vote on [this discussion](https://github.com/openshift/windows-machine-config-operator/pull/1284/commits/13e9b4c8f0d13577a9ad574ed20d0bc7f9b566c6#r1012929387) is:

- I'm a proponent for minimizing platform-specific logic. Therefore, I am fine if we always transfer the `gcp-get-valid-hostname.ps1` script as part of the payload (the approach proposed in the second commit [ee0c52e](https://github.com/openshift/windows-machine-config-operator/pull/1284/commits/ee0c52e372f54ed69b004ef719aae576a9e5ef9f#diff-4d8fe3bb66a872f0e4a277f820d60d2efa722f09aa89516be74b98c06b57f3bdR155))

File: pkg/services/services.go (hunk 2 above)
Comment (jrvaldes) [reply pkg/services/services.go]: Makes sense, will squash the commits. 

Thanks you all for the feedback.

File: pkg/services/services.go (hunk 4 above)
Comment (jrvaldes) [reply pkg/services/services.go]: The proposed implementation will take care of this so that the naming is consistent, even if the length of the instance's name is greater than 63 chars.

File: pkg/services/services.go (hunk 6 above)
Comment (jrvaldes) [reply pkg/services/services.go]: Addressed.

File: pkg/services/services.go (hunk 3 above)
Comment (jrvaldes) [reply pkg/services/services.go]: >  a reference to a GCE doc that guarantees this endpoint won't change

Couldn't find such a statement explicitly in the docs or source code.

> way for us to easily catch the change in an e2e test.

Sure, it can be added in case we decided to consume the endpoint directly.


As mentioned, using the Cloud Tools for PowerShell library provides a safe abstraction from maintaining the instance metadata endpoint as the library is installed in the instance by the cloud provider.

Back to https://github.com/openshift/windows-machine-config-operator/pull/1260#discussion_r997394197  about the  BYOH use case, it may be an issue if the instance does not have the library. I suggest the following options:
 - Add the `Cloud Tools for PowerShell` library as a BYOH requirement for AWS and GCP (at the moment), or
 - Invoke the instance metadata endpoint directly to eliminate the dependency on the  `Cloud Tools for PowerShell` library

Comment (aravindhp) [reply pkg/services/services.go]: > Couldn't find such a statement explicitly in the docs or source code.

[Default VM metadata values](https://cloud.google.com/compute/docs/metadata/default-metadata-values) should suffice.

> Invoke the instance metadata endpoint directly to eliminate the dependency on the Cloud Tools for PowerShell library

I am not against this approach and it is the best path forward.

File: pkg/internal/gcp-get-valid-hostname.ps1 [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# This script returns a mutated hostname if the length is longer than 63 characters. The hostname will be the
```
Comment (aravindhp) [inline pkg/internal/gcp-get-valid-hostname.ps1:1]: `/gcp-get-hostname` will suffice. Valid is implied and obvious. If the script is returning an invalid hostname then it is a bug.

File: pkg/internal/gcp-get-valid-hostname.ps1 [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+# lesser of 63 characters after the first dot in the FQDN. Uses the GoogleCloud Tools for PowerShell (https://cloud.google.com/tools/powershell/docs/quickstart)
+# to get the original hostname from the GCP instance metadata service.
+
+# MAX_LENGTH is the maximum number of character allowed for the instance's hostname in GCP
+$MAX_LENGTH = 63
+
+# get hostname from the instance metadata service
+$hostname=Get-GceMetadata -Path "instance/hostname"
```
Comment (aravindhp) [inline pkg/internal/gcp-get-valid-hostname.ps1:9]: As discussed it is best to go the internal endpoint due to the BYOH case you have pointed out. Please link to the doc I pointed you at.

File: pkg/services/services.go [hunk 9]
```diff
@@ -144,13 +143,11 @@ func getKubeletServiceConfiguration(argsFromIginition map[string]string, debug b
 		return servicescm.Service{}, err
 	}
 	var powershellVars []servicescm.PowershellCmdArg
-	hostnameArg := getKubeletHostnameOverride(platform)
-	if hostnameArg != "" {
-		kubeletArgs = append(kubeletArgs, hostnameArg)
-		powershellVars = append(powershellVars, servicescm.PowershellCmdArg{
-			Name: ec2HostnameVar,
-			Path: "Get-EC2InstanceMetadata -Category LocalHostname",
```
Comment (aravindhp) [inline pkg/services/services.go:152]: This is the only thing that changes across cloud providers, so your helper should just focus on that.

File: pkg/services/services.go [hunk 10]
```diff
@@ -218,11 +215,34 @@ func klogVerbosityArg(debug bool) string {
...
+		return "", servicescm.PowershellCmdArg{}
+	}
+}
+
+// getAWSMetadataHostnamePowershellCmdArg returns the PowershellCmdArg to resolve the hostname using the instance
+// metadata service in AWS. Uses AWS Tools for PowerShell (https://docs.aws.amazon.com/powershell/latest/userguide/pstools-welcome.html)
+func getAWSMetadataHostnamePowershellCmdArg(variableName string) servicescm.PowershellCmdArg {
+	return servicescm.PowershellCmdArg{
+		Name: variableName,
+		Path: "Get-EC2InstanceMetadata -Category LocalHostname",
```
Comment (aravindhp) [inline pkg/services/services.go:236]: The helper should only deal with the command not the whole `PowershellCmdArg` object. 
`GetAWSHostnameCmd` is succinct and sufficiently informs us what the helper does.

File: pkg/services/services.go [hunk 11]
```diff
@@ -218,11 +215,34 @@ func klogVerbosityArg(debug bool) string {
...
+	}
+}
+
+// getGCPMetadataHostnamePowershellCmdArg returns the PowershellCmdArg to resolve the hostname using the instance
+// metadata service in GCP.
+// GCP is a special case, and the resulting hostname is truncated if the length is longer than 63 characters.
+func getGCPMetadataHostnamePowershellCmdArg(variableName string) servicescm.PowershellCmdArg {
```
Comment (aravindhp) [inline pkg/services/services.go:243]: `GetGCPHostnameCmd`

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/internal/gcp-get-valid-hostname.ps1 (hunk 7 above)
Comment (saifshaikh48) [reply pkg/internal/gcp-get-valid-hostname.ps1:1]: +1, applies throughout PR

File: pkg/services/services.go (hunk 10 above)
Comment (jrvaldes) [reply pkg/services/services.go]: Makes sense.

File: pkg/windows/windows.go [hunk 12]
```diff
@@ -25,6 +25,8 @@ const (
 	remoteDir = "C:\\Temp\\"
 	// winTemp is the default Windows temporary directory
 	winTemp = "C:\\Windows\\Temp\\"
+	// GcpGetValidHostnameScriptRemotePath a
```
[the remark this answers is not in the document]
Comment (jrvaldes) [reply pkg/windows/windows.go]: Fixed.

File: test/e2e/create_test.go [hunk 13]
```diff
@@ -589,3 +593,46 @@ func (tc *testContext) scaleDeployment(namespace, name, selector string, expecte
...
 	return nil
 }
+
+// checkMachineInstanceMetadata checks the instance metadata on the Windows machines and returns with success
+// after finding the first machine where the instance metadata endpoint is available. Otherwise, returns an error.
+// This check is platform specific, as the instance metadata endpoint is required in a subset of the supported
+// platforms (i.e. GCP).
+// The BYOH use case is also covered by this check, since both machine and BYOH instances runs on the same cloud
+// provider.
+func (tc *testContext) checkMachineInstanceMetadata(machines []mapi.Machine) (err error) {
```
Comment (sebsoto) [inline test/e2e/create_test.go:603]: I dont see the value in this helper.
Its not testing the functionality of our code.
If the metadata endpoint isnt reachable, WMCO will fail to configure the VM.

File: pkg/services/services_test.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetHostnameCmd(t *testing.T) {
+	tests := []struct {
+		name         string
+		platformType config.PlatformType
+		expected     string
+	}{
+		// TODO: Add test cases.
```
Comment (sebsoto) [inline pkg/services/services_test.go:16]: This TODO isn't helpful as it doesn't describe what test cases are missing.
Additionally if you know what test cases are missing that are required, please just add them

**Review by sebsoto — CHANGES_REQUESTED** [review]
Thanks @jrvaldes PTAL at my comments

File: pkg/services/services_test.go (hunk 14 above)
Comment (jrvaldes) [reply pkg/services/services_test.go:16]: Missed removing this line. Good catch.

File: test/e2e/create_test.go (hunk 13 above)
Comment (jrvaldes) [reply test/e2e/create_test.go:603]: Correct. This test is redundant as WMCO won't be able to configure the Windows instance if the address of the instance metadata endpoint changes. 

Removed the test.

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

:rocket:

**Comment by alinaryan** [comment]
LGTM

File: test/e2e/create_test.go (hunk 13 above)
Comment (aravindhp) [reply test/e2e/create_test.go]: I am fine with not having a test as long as WMCO failing to configuring the VM will point to an issue with the endpoint.

File: pkg/internal/gcp-get-hostname.ps1 [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+#
+# The IPv4 address of the instance metadata service is defaulted to 169.254.169.254. See instance metadata
+# documentaion for GCP https://cloud.google.com/compute/docs/metadata/default-metadata-values.
+
+# MAX_LENGTH is the maximum number of character allowed for the instance's hostname in GCP
+$MAX_LENGTH = 63
+
+# get hostname from the instance metadata service
+$hostname=(Invoke-RestMethod -Headers @{'Metadata-Flavor'='Google'} `
+            -Uri "http://169.254.169.254/computeMetadata/v1/instance/hostname")
```
Comment (aravindhp) [inline pkg/internal/gcp-get-hostname.ps1:17]: Going with the doc this should be http://metadata.google.internal and not the IP address.

File: pkg/services/services.go [hunk 16]
```diff
@@ -19,10 +19,9 @@ const (
 	// https://docs.openshift.com/container-platform/latest/rest_api/editing-kubelet-log-level-verbosity.html#log-verbosity-descriptions_editing-kubelet-log-level-verbosity
 	debugLogLevel    = "4"
 	standardLogLevel = "2"
-	// ec2HostnameVar is a variable that should be replaced with the value of `local-hostname` in the ec2 instance
-	// metadata
-	ec2HostnameVar = "EC2_HOSTNAME"
-	NodeIPVar      = "NODE_IP"
+	// hostnameOverrideVarName is a variable that should be replaced with the value of the desired instance hostname
+	hostnameOverrideVarName = "HOSTNAME_OVERRIDE"
```
Comment (aravindhp) [inline pkg/services/services.go:23]: This should be `hostNameOverrideVar`

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/internal/gcp-get-hostname.ps1 (hunk 15 above)
Comment (jrvaldes) [reply pkg/internal/gcp-get-hostname.ps1:17]: There is a [comment](https://github.com/openshift/windows-machine-config-operator/pull/1284/files#diff-e5bc5d4c1a111c9f29a32845d5d34069f73857555f955fe898efeed077cdcba7R4) up in this file with details about this. 

TL;DR
Using the IP address is safer, especially in the BYOH use case where there is no control over the host file and the resolution of the  `metadata.google.internal` DNS name may not be available.

Comment (aravindhp) [reply pkg/internal/gcp-get-hostname.ps1:17]: Won't images be created from existing master images which will have this DNS setup? Now we could run into issues if the IP has changed. Or is there some way to import images created on prem where this need not have setup?

**Review by aravindhp — APPROVED** [review]

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

nice work @jrvaldes

**Comment by aravindhp** [comment]
@jrvaldes please don't manually cherry-pick this. This should apply cleanly to 4.12 and most likely is missing some other backport.

**Comment by jrvaldes** [comment]
> @jrvaldes please don't manually cherry-pick this. This should apply cleanly to 4.12 and most likely is missing some other backport.

I don't plan to. https://github.com/openshift/windows-machine-config-operator/pull/1314 must go first.
